File: src/rust/planning_poker/src/usecases/change_user_mode.rs
use crate::{
    domains::game_player::{
        GamePlayer, GamePlayerId, GamePlayerRepository, HaveGamePlayerRepository, UserMode,
    },
    utils::types::LocalBoxFuture,
};

#[cfg(test)]
mod tests;

pub trait ChangeUserModeDependency: HaveGamePlayerRepository {}

#[derive(PartialEq, Debug)]
pub enum ChangeUserModeOutput {
    NotFound,
}

pub trait ChangeUserMode {
    fn execute(
        &self,
        game_player_id: GamePlayerId,
        mode: UserMode,
    ) -> LocalBoxFuture<'_, Result<GamePlayer, ChangeUserModeOutput>>;
}

impl<T: ChangeUserModeDependency> ChangeUserMode for T {
    fn execute(
        &self,
        game_player_id: GamePlayerId,
        mode: UserMode,
    ) -> LocalBoxFuture<'_, Result<GamePlayer, ChangeUserModeOutput>> {
        let repository = self.get_game_player_repository();

        let fut = async move {
            let user = repository.find_by(game_player_id).await;

            match user {
                None => Err(ChangeUserModeOutput::NotFound),
                Some(mut player) => {
                    player.change_user_mode(mode, |_| {});
                    repository.save(&player).await;

                    Ok(player)
                }
            }
        };

        Box::pin(fut)
    }
}

File: src/usecases/change-user-mode.ts
import * as GamePlayer from "@/domains/game-player";
import * as User from "@/domains/user";
import * as Game from "@/domains/game";
import { GameRepository } from "@/domains/game-repository";
import { UseCase } from "./base";

export interface ChangeUserModeInput {
  userId: User.Id;
  gameId: Game.Id;
  mode: GamePlayer.UserMode;
}

export type ChangeUserModeOutput = { kind: "success"; game: Game.T } | { kind: "notFound" };

export class ChangeUserModeUseCase implements UseCase<ChangeUserModeInput, Promise<ChangeUserModeOutput>> {
  constructor(private gameRepository: GameRepository) {}

  async execute(input: ChangeUserModeInput): Promise<ChangeUserModeOutput> {
    const game = await this.gameRepository.findBy(input.gameId);
    if (!game) {
      return { kind: "notFound" };
    }

    const newGame = Game.declarePlayerTo(game, input.userId, input.mode);
    await this.gameRepository.save(newGame);

    return { kind: "success", game: newGame };
  }
}
